openstack/networking/v2/extensions/qos/results.go

```diff
@@ -0,0 +1,32 @@
+package qos
+
+import "github.com/gophercloud/gophercloud"
+
+// The result of listing the qos rule types
+type ListRuleTypesResult struct {
+	gophercloud.Result
+}
+
+func (l ListRuleTypesResult) Extract() (out []string, err error) {
+	// the response from the server is in this format, but it's not an especially useful format
+	// so we will return the rule types, in a slice of strings
+	var list struct {
+		RuleTypes []struct {
+			Type string `json:"type"`
+		} `json:"rule_types"`
+	}
+
+	// take the JSON and put it into the list struct
+	if err = l.ExtractInto(&list); err != nil {
+		return
+	}
+
+	// now make the out slice
+	out = make([]string, len(list.RuleTypes))
+	// and copy the result to the return value
+	for i, t := range list.RuleTypes {
+		out[i] = t.Type
+	}
+
+	return
+}
```

Comment:
I've debated this one for almost an hour. I understand that the results of retrieving the details of a rule type are very different than the list request and that the list request is only returning a `Type` field, but I think we should create a proper struct for `RuleType`, even though it'll only contain one field.

The more detailed result can be called `RuleTypeDetails`. Or choose names you feel are suitable.

I realize that this basically rewrites this entire file and I apologize about that. But I have a feeling this is better in the long-run. To help get you started, the contents of the file would look something like:

```go
type RuleType struct {
  Type string `json:"type"`
}

// The result of listing the qos rule types
type ListRuleTypesPage struct {
  pagination.SinglePageBase
}

// IsEmpty indicates whether a RuleTypesPage is empty.
func (r ListRuleTypesPage) IsEmpty() (bool, error) {
  v, err := ExtractRuleTypes(r)
  return len(v) == 0, err
}

func ExtractRuleTypes(r pagination.Page) ([]RuleType, error) {
  var s struct {
    RuleTypes []RuleType `json:"rule_types"`
  }
  err := (r.(ListRuleTypesPage)).ExtractInto(&s)
  return s.RuleTypes, err
}
```